### PR TITLE
[ROCm][DSV4][Perf] Fuse inverse-RoPE and cache bf16 wo_a in o-projection

### PR DESCRIPTION
**Pull request by Fangzhou-Ai**
## Purpose
The ROCm DeepSeek V4 o-projection (rocm_inv_rope_einsum, run once per decode layer) ran the inverse GPT-J RoPE as ~10 small PyTorch kernels  (clone, index_select, 2x repeat_interleave, neg, stack, cat, casts) and re-dequantized the static fp8 wo_a weight every step (fp8->fp32->*scale ->bf16), which showed up in the decode profile as the two largest copy/mul kernels.
    - Add `_inverse_rope_gptj_kernel` (Triton): single-launch inverse GPT-J RoPE writing bf16 directly. Guarded for the non-neox, even-rope_dim, contiguous layout; the PyTorch path is kept as fallback.
    - Cache the dequantized bf16 wo_a weight on the module  (`_dsv4_wo_a_bf16`) so the dequant runs once instead of per step.

## Test Result
Unit test result shows bitwise match.

ISL/OSL 8k1k tests:
Concurrency | Baseline tok/s | New tok/s | Δ tok/s | Throughput gain | Baseline mean TPOT | New mean TPOT | Δ TPOT | TPOT reduction
-- | -- | -- | -- | -- | -- | -- | -- | --
4 | 78.03 | 90.19 | +12.16 | +15.6% | 48.57 ms | 42.10 ms | -6.47 ms | 13.3% better
8 | 141.18 | 162.67 | +21.49 | +15.2% | 52.77 ms | 45.75 ms | -7.02 ms | 13.3% better
16 | 238.72 | 271.56 | +32.84 | +13.8% | 61.43 ms | 53.96 ms | -7.47 ms | 12.2% better
32 | 390.06 | 434.76 | +44.70 | +11.5% | 74.92 ms | 67.29 ms | -7.63 ms | 10.2% better
64 | 578.68 | 630.30 | +51.62 | +8.9% | 102.52 ms | 94.37 ms | -8.15 ms | 7.9% better

GSM8K test:
Task | Version | Filter | n-shot | Metric | Value | Stderr
-- | -- | -- | -- | -- | -- | --
gsm8k | 3 | flexible-extract | 20 | exact_match | 0.9583 | ±0.0055
gsm8k | 3 | strict-match | 20 | exact_match | 0.9583 | ±0.0055

### CODE REVIEW

**Comment by dllehr-amd** [comment]
hey @Fangzhou-Ai  can you clear the DCO error, and sign your commits on this PR?!

File: tests/kernels/attention/test_rocm_triton_attn_dsv4.py [hunk 1]
```diff
@@ -375,3 +375,305 @@ def test_combine_topk_swa_indices_ragged() -> None:
...
+def _make_cos_sin_cache(
+    max_pos: int, rope_dim: int, device: torch.device
+) -> torch.Tensor:
+    """fp32 cos_sin_cache laid out as [P, rope_dim] = cos[:half] | sin[half:]."""
+    half = rope_dim // 2
+    angles = torch.rand(max_pos, half, dtype=torch.float32, device=device) * 6.2831853
+    return torch.cat((angles.cos(), angles.sin()), dim=-1)
+
+
+class _FakeGPTJRotary(torch.nn.Module):
```
Comment (dllehr-amd) [inline tests/kernels/attention/test_rocm_triton_attn_dsv4.py]: I have a small question.  in the UT we're driving the fake rotary.  And the `test_rocm_inv_rope_einsum`  tests, both fused, and fallback, uses the fake rotary.  But we never go against the `DeepseekV4ScalingRotaryEmbedding`  as you commented here.  Would it be worth testing the official DeepseekV4ScalingRotaryEmbedding?  My only concern is if the DeepseekV4ScalingRotaryEmbedding  changes in the future, this test may become inadequate? 

What're your thoughts?

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -951,11 +1020,42 @@ def rocm_inv_rope_einsum(
...
+    n_local_groups: int,
+    o_lora_rank: int,
+    wo_a: torch.nn.Module,
+) -> torch.Tensor:
+    """Inverse-RoPE + WO_A bmm path used on ROCm.
+
+    Fuses the inverse GPT-J RoPE into one Triton kernel and caches the bf16
+    wo_a weight so the per-step dequant disappears.
+    """
+    if _can_use_fused_inv_rope(rotary_emb, o, rope_head_dim):
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: `rocm_inv_rope_einsum` is only used for DSv4 and `_can_use_fused_inv_rope(rotary_emb, o, rope_head_dim)` is always `True` for both Flash and Pro models

We should just replace `_apply_inv_rope_ref`

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 3]
```diff
@@ -924,22 +893,122 @@ def _apply_inv_rope_ref(
...
+    b = tl.load(o_ptr + in_base + NOPE + 2 * k + 1, mask=kmask).to(tl.float32)
+    cos = tl.load(cos_sin_ptr + pos * cs_stride + k, mask=kmask)
+    sin = tl.load(cos_sin_ptr + pos * cs_stride + HALF + k, mask=kmask)
+    out_even = a * cos + b * sin
+    out_odd = b * cos - a * sin
+    tl.store(out_ptr + out_base + NOPE + 2 * k, out_even.to(tl.bfloat16), mask=kmask)
+    tl.store(out_ptr + out_base + NOPE + 2 * k + 1, out_odd.to(tl.bfloat16), mask=kmask)
+
+
+def _can_use_fused_inv_rope(
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: We don't need this function anymore since it is always evaluated to true for DeepSeekV4 Models.

Comment (tjtanaa) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: After making these changes please also update the unit tests as I saw the unittest is also using this helper function.

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py (hunk 2 above)
Comment (Fangzhou-Ai) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: helper removed

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py (hunk 3 above)
Comment (Fangzhou-Ai) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: helper removed

File: tests/kernels/attention/test_rocm_triton_attn_dsv4.py (hunk 1 above)
Comment (Fangzhou-Ai) [reply tests/kernels/attention/test_rocm_triton_attn_dsv4.py]: I agree, the updates have been made to instantiate a real official DeepseekV4ScalingRotaryEmbedding and compare the fused kernel directly against forward_native(inverse=True).

**Review by dllehr-amd — APPROVED** [review]
I'm happy with this change as it is now.  Approving.  Thanks @Fangzhou-Ai
